misc/vimiv.1
.\" Man page generated from reStructuredText.
.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.TH "VIMIV" "1" "Aug 26, 2023" "" "vimiv"
.SH NAME
vimiv \- an image viewer with vim-like keybindings
.SH SYNOPSIS
.sp
\fBvimiv\fP [\fBPATH\fP] [\fB\-\-help\fP|\fB\-h\fP] [\fB\-\-fullscreen\fP|\fB\-f\fP] [\fB\-\-version\fP|\fB\-v\fP] [\fB\-\-geometry\fP \fIWIDTHxHEIGHT\fP|\fB\-g\fP \fIWIDTHxHEIGHT\fP] [\fB\-\-temp\-basedir\fP] [\fB\-\-config\fP \fIFILE\fP] [\fB\-\-keyfile\fP \fIFILE\fP] [\fB\-\-set\fP \fIOPTION\fP \fIVALUE\fP|\fB\-s\fP \fIOPTION\fP \fIVALUE\fP] [\fB\-\-log\-level\fP \fILEVEL\fP] [\fB\-\-command\fP \fICOMMAND\fP] [\fB\-\-basedir\fP \fIDIRECTORY\fP|\fB\-b\fP \fIDIRECTORY\fP] [\fB\-\-output\fP \fITEXT\fP|\fB\-o\fP \fITEXT\fP] [\fB\-\-input\fP|\fB\-i\fP] [\fB\-\fP] [\fB\-\-debug\fP \fIMODULE\fP] [\fB\-\-qt\-args\fP \fIARGS\fP]
.SH DESCRIPTION
.sp
Vimiv is an image viewer with vim\-like keybindings. It is written in python3
using the Qt5 toolkit and is free software, licensed under the GPL. Some of the
features are:
.INDENT 0.0
.IP \(bu 2
Simple library browser
.IP \(bu 2
Thumbnail mode
.IP \(bu 2
Basic image editing
.IP \(bu 2
Command line with tab completion
.IP \(bu 2
Complete customization with style sheets
.UNINDENT
.sp
A much more complete documentation can be found under
\fI\%https://karlch.github.io/vimiv\-qt/\fP\&.
.SH OPTIONS
.SS POSITIONAL ARGUMENTS
.sp
\fBPATH\fP
.INDENT 0.0
.INDENT 3.5
Paths to open
.UNINDENT
.UNINDENT
.SS OPTIONS
.sp
\fB\-\-help\fP, \fB\-h\fP
.INDENT 0.0
.INDENT 3.5
show this help message and exit
.UNINDENT
.UNINDENT
.sp
\fB\-\-fullscreen\fP, \fB\-f\fP
.INDENT 0.0
.INDENT 3.5
Start fullscreen
.UNINDENT
.UNINDENT
.sp
\fB\-\-version\fP, \fB\-v\fP
.INDENT 0.0
.INDENT 3.5
Print version information and exit
.UNINDENT
.UNINDENT
.sp
\fB\-\-geometry\fP \fIWIDTHxHEIGHT\fP, \fB\-g\fP \fIWIDTHxHEIGHT\fP
.INDENT 0.0
.INDENT 3.5
Set the starting geometry
.UNINDENT
.UNINDENT
.sp
\fB\-\-temp\-basedir\fP
.INDENT 0.0
.INDENT 3.5
Use a temporary basedir
.UNINDENT
.UNINDENT
.sp
\fB\-\-config\fP \fIFILE\fP
.INDENT 0.0
.INDENT 3.5
Use FILE as local configuration file
.UNINDENT
.UNINDENT
.sp
\fB\-\-keyfile\fP \fIFILE\fP
.INDENT 0.0
.INDENT 3.5
Use FILE as keybinding file
.UNINDENT
.UNINDENT
.sp
\fB\-\-set\fP \fIOPTION\fP \fIVALUE\fP, \fB\-s\fP \fIOPTION\fP \fIVALUE\fP
.INDENT 0.0
.INDENT 3.5
Set a temporary setting
.UNINDENT
.UNINDENT
.sp
\fB\-\-log\-level\fP \fILEVEL\fP
.INDENT 0.0
.INDENT 3.5
Set log level to LEVEL
.UNINDENT
.UNINDENT
.sp
\fB\-\-command\fP \fICOMMAND\fP
.INDENT 0.0
.INDENT 3.5
Run COMMAND on startup, usable multiple times
.UNINDENT
.UNINDENT
.sp
\fB\-\-basedir\fP \fIDIRECTORY\fP, \fB\-b\fP \fIDIRECTORY\fP
.INDENT 0.0
.INDENT 3.5
Directory to use for all storage
.UNINDENT
.UNINDENT
.sp
\fB\-\-output\fP \fITEXT\fP, \fB\-o\fP \fITEXT\fP
.INDENT 0.0
.INDENT 3.5
Wildcard expanded string to print to standard output upon quit
.UNINDENT
.UNINDENT
.sp
\fB\-\-input\fP, \fB\-i\fP
.INDENT 0.0
.INDENT 3.5
Read paths to open from standard input
.UNINDENT
.UNINDENT
.sp
\fB\-\fP
.INDENT 0.0
.INDENT 3.5
Read binary image to open from standard input
.UNINDENT
.UNINDENT
.SS DEVELOPMENT ARGUMENTS
.sp
\fB\-\-debug\fP \fIMODULE\fP
.INDENT 0.0
.INDENT 3.5
Force showing debug log messages of MODULE
.UNINDENT
.UNINDENT
.sp
\fB\-\-qt\-args\fP \fIARGS\fP
.INDENT 0.0
.INDENT 3.5
Arguments to pass to qt directly, use quotes to pass multiple arguments
.UNINDENT
.UNINDENT
.SH WEBSITE
.sp
\fI\%https://karlch.github.io/vimiv\-qt/\fP
.SH AUTHOR
vimiv was designed and written by Christian Karl. All contributors are listed in the AUTHORS file distributed with vimiv.
.SH COPYRIGHT
The vimiv contributors, licensed under the GNU GPL v3
.\" Generated by docutils manpage writer.
.


docs/changelog.rst
Changelog
=========

All notable changes to vimiv are documented in this file.

v0.10.0 (unreleased)
--------------------

Added:
^^^^^^

* Support for PyQt6 and thus Qt6. In case both are installed, vimiv prefers PyQt6 and
  only falls back to PyQt5 if PyQt6 is not available. The Qt version and wrapper can be
  selected explicitly with the ``QT_SELECT`` environment variable, using:

  * ``5`` or ``PyQt5``: Use PyQt5.
  * ``6`` or ``PyQt6``: Use PyQt6.
  * ``PySide6``: Use PySide6 (Qt for Python). This is highly experimental and should be
    used with care.

Changed:
^^^^^^^^

* The JPEG image header check was simplified to have a false negative rate of 0, while
  maintaining a decently low false positive rate.
* Support for PyQt5 < 5.15 was dropped.

Fixed:
^^^^^^

* The SVG image header check to no longer return true for all XML files.
* Binding the ``<delete>`` key as special key. Thanks `@xfzv`_!
* Consecutive `:tag-write` would insert empty lines into the tag file, which would on
  `:tag-load` getting interpreted as file paths.


v0.9.0 (2023-07-15)
-------------------

Added:
^^^^^^

* More vim movement commands with corresponding keybindings, namely:

  * ``:scroll page-up/page-down/half-page-up/half-page-down`` mapped to ``<ctrl>b``,
    ``<ctrl>f``, ``<ctrl>u`` and ``<ctrl>d`` for library and thumbnail mode.
  * ``:end-of-line`` mapped to ``$`` and ``:first-of-line`` mapped to ``^`` to select
    the first icon of the current row in thumbnail mode.

  Thanks `@jcjgraf <https://github.com/jcjgraf>`_ for the idea!
* The ``--action`` flag for the ``:mark`` command. ``--action toggle`` is the default
  and keeps the previous behaviour of the mark command, i.e. inversing the mark status
  of the passed path(s). In addition one can now use ``--action mark`` and
  ``--action unmark`` to force (un-) marking the paths.
* Support for the ``avif`` file format using the imageformats plugin. To enable it, add
  ``imageformats = avif`` to the ``[PLUGINS]`` section of your ``vimiv.conf``. Requires
  `qt-avif-image-plugin <https://github.com/novomesk/qt-avif-image-plugin>`_.
  Thanks `@schyzophrene-asynchrone`_!
* A few more default mouse bindings:

  * ``<button-right>`` to enter the library from image and thumbnail mode.
  * ``<button-right>`` to go to the parent directory in the library.
  * ``<button-middle>`` to enter thumbnail mode from image and library mode.
  * ``<button-forward>`` and ``<button-back>`` to select the next/previous image in
    image mode.
  * ``<button-forward>`` and ``<button-back>`` to scroll down/up and open any selected
    image in library mode.
  * ``<button-forward>`` and ``<button-back>`` select the next/previous thumbnail in
    thumnbail mode.

  Thanks `@BachoSeven`_ for your thoughts!
* The ``:copy-image`` command which copies the selected image to the clipboard. The
  flags ``--width``, ``--height`` and ``--size`` allow to scale the copied image.
  Thanks `@jcjgraf`_!
* The ``read_only`` setting. When set to true, all commands that may write changes to
  disk are disabled. This comes with the ``{read-only}`` status module which displays
  ``[RO]`` in case ``read_only`` is true. Note that you must add this to your
  vimiv.conf to enable it if you already define the left status contents. The new
  defaults are::

    left = {pwd}{read-only}
    left_image = {index}/{total} {basename}{read-only} [{zoomlevel}]
    left_thumbnail = {thumbnail-index}/{thumbnail-total} {thumbnail-name}{read-only}

  Thanks `@willemw12`_ for the idea!
* The ``zh`` keybinding to toggle hidden files.
  Thanks `@Kakupakat`_ for the idea!
* The ``:print-stdout`` command which prints a given list of string to the STDOUT.
* The ``:mark-print`` default alias which prints all marked images to STDOUT using
  ``:print-stdout``.
* The ``-o TEXT`` flag to print ``TEXT`` to standard output before exit. Wildcards are
  expanded before printing, allowing to for example print marked files using ``-o %m``.
  Thanks `@loiccoyle`_ for the idea!
* The ``-i`` flag to read paths to open from standard input instead.
  Thanks `@loiccoyle`_ for the idea!
* The ``--qt-args`` flag to pass arguments directly to qt. When passing multiple
  arguments, they must be quoted accordingly, e.g.
  ``vimiv --qt-args '--name floating'``.
  Thanks `@loiccoyle`_ for the discussion!
* Statusbar modules ``name``, ``thumbnail-basename``, ``extension`` and
  ``thumbnail-extension``.
* Support for the ``jp2`` file format using the imageformats plugin. To enable it, add
  ``imageformats = jp2`` to the ``[PLUGINS]`` section of your ``vimiv.conf``. Requires
  the qt imageformats plugin. Thanks `@szsdk`_ for testing!
* The ``image.zoom_wheel_ctrl`` setting which toggles the need to hold the ``<ctrl>``
  modifier for zooming an image with the mouse wheel. Thanks `@ArtemSmaznov`_ for the
  idea!
* The ``--ask`` flag for the ``:delete`` command to prompt the user for confirmation
  before deleting. Thanks `@timsofteng`_ for the idea!
* Support for different sorting options for images and directories via the
  ``sort.image_order``, ``sort.directory_order``, ``sort.reverse`` and
  ``sort.ignore_case`` settings. Please refer to the :ref:`documentation <sorting>` for
  further details. Thanks to `@kAldown`_ for the initial implementation, `@jcjgraf`_ for
  taking over, and many more joining in the discussions and reminding us why this
  feature is important!
* New ``{cursor-position}`` statusbar module which tracks the mouse cursor position in
  image coordinates.
* The ``:crop`` command which displays a rectangle to crop the curent image. The
  rectangle can be dragged and resized using the mouse. As with ``:straighten``, accept
  the changes with ``<return>`` and reject them with ``<escape>``. The
  ``{transformation-info}`` status module displays the currently selected geometry of
  the original image. Thanks `@Yutsuten`_ for reviving this!
* Add the ``none`` sorting type for the ``sort.image_order`` and ``sort.directory_order``

vimiv/vimiv.py
# vim: ft=python fileencoding=utf-8 sw=4 et sts=4

# This file is part of vimiv.
# Copyright 2017-2019 Christian Karl (karlch) <karlch at protonmail dot com>
# License: GNU GPL v3, see the "LICENSE" and "AUTHORS" files for details.

"""Main function and startup utilities for vimiv.

Module Attributes:
    _tmpdir: TemporaryDirectory when running with ``--temp-basedir``. The
        object must exist until vimiv exits.
"""

import argparse
import logging
import logging.handlers
import os
import signal
import sys
import tempfile

from PyQt5.QtWidgets import QApplication

import vimiv
from vimiv import app, api, parsertypes, imutils, plugins
from vimiv.completion import completionmodels
from vimiv.config import configfile, keyfile, styles
from vimiv.gui import mainwindow
from vimiv.utils import (
    xdg,
    clipboard,
    statusbar_loghandler,
    strconvert,
    trash_manager,
    working_directory,
)


_tmpdir = None


def startup(argv):
    """Run the functions to set up everything.

    Args:
        argv: sys.argv[1:] from the executable.
    """
    args = get_argparser().parse_args(argv)
    init_directories(args)
    setup_logging(args.log_level)
    logging.debug("Start: vimiv %s", " ".join(argv))
    update_settings(args)
    earlyinit()
    init_ui(args)
    plugins.load()
    init_paths(args)
    logging.debug("Startup completed")
    api.status.update()


def earlyinit():
    """Initialize objects needed as early as possible."""
    logging.debug("Initializing early objects")
    clipboard.init()
    trash_manager.init()
    working_directory.init()
    imutils.init()
    completionmodels.init()


def setup_logging(log_level):
    """Prepare the python logging module.

    Sets it up to write to stderr and $XDG_DATA_HOME/vimiv/vimiv.log.

    Args:
        log_level: Log level as string as given from the command line.
    """
    log_format = logging.Formatter(
        "[%(asctime)s] %(levelname)s: %(message)s", datefmt="%H:%M:%S"
    )

    logger = logging.getLogger()
    logger.handlers = []
    logger.setLevel(log_level)

    file_handler = logging.FileHandler(xdg.join_vimiv_data("vimiv.log"), mode="w")
    file_handler.setFormatter(log_format)
    logger.addHandler(file_handler)

    console_handler = logging.StreamHandler()
    console_handler.setFormatter(log_format)
    logger.addHandler(console_handler)

    sb_handler = statusbar_loghandler.StatusbarLogHandler()
    logger.addHandler(sb_handler)


def get_argparser():
    """Get the argparse parser."""
    parser = argparse.ArgumentParser(
        prog=vimiv.__name__, description=vimiv.__description__
    )
    parser.add_argument(
        "-f", "--fullscreen", action="store_true", help="Start fullscreen"
    )
    parser.add_argument(
        "-v",
        "--version",
        action="version",
        version="vimiv version %s" % vimiv.__version__,
        help="Print version information and exit",
    )
    parser.add_argument(
        "--slideshow", action="store_true", help="Start slideshow at start-up"
    )
    parser.add_argument(
        "-g",
        "--geometry",
        type=parsertypes.geometry,
        metavar="WIDTHxHEIGHT",
        help="Set the starting geometry",
    )
    parser.add_argument(
        "--temp-basedir", action="store_true", help="Use a temporary basedir"
    )
    parser.add_argument(
        "--config",
        type=parsertypes.existing_file,
        metavar="FILE",
        help="Use FILE as local configuration file",
    )
    parser.add_argument(
        "--keyfile",
        type=parsertypes.existing_file,
        metavar="FILE",
        help="Use FILE as keybinding file",
    )
    parser.add_argument(
        "-s",
        "--set",
        nargs=2,
        default=[],
        action="append",
        dest="cmd_settings",
        metavar=("OPTION", "VALUE"),
        help="Set a temporary setting",
    )
    parser.add_argument(
        "--log-level",
        type=parsertypes.loglevel,
        metavar="LEVEL",
        help="Set log level to LEVEL",
        default="info",
    )
    parser.add_argument(
        "paths",
        nargs="*",
        type=parsertypes.existing_path,
        metavar="PATH",
        help="Paths to open",
    )
    return parser


def init_directories(args):
    """Create vimiv cache, config and data directories.

    The directories are either the directories defined in the freedesktop
    standard or located in a temporary base directory.

    Args:
        args: Arguments returned from parser.parse_args().
    """
    if args.temp_basedir:
        global _tmpdir
        _tmpdir = tempfile.TemporaryDirectory(prefix="vimiv-tempdir-")
        basedir = _tmpdir.name
        os.environ["XDG_CACHE_HOME"] = os.path.join(basedir, "cache")
        os.environ["XDG_CONFIG_HOME"] = os.path.join(basedir, "config")
        os.environ["XDG_DATA_HOME"] = os.path.join(basedir, "data")
    for directory in [
        xdg.vimiv_cache_dir(),
        xdg.vimiv_config_dir(),
        xdg.vimiv_data_dir(),
        xdg.join_vimiv_config("styles"),
    ]:
        os.makedirs(directory, exist_ok=True)


def init_paths(args):
    """Open paths given from commandline or fallback to library if set."""
    logging.debug("Opening paths")
    if not app.open_paths(args.paths) and api.settings.STARTUP_LIBRARY.value:
        app.open_paths([os.getcwd()])


def init_ui(args):
    """Initialize the Qt UI."""
    logging.debug("Initializing UI")
    mw = mainwindow.MainWindow()
    mw.show()
    if args.fullscreen:
        mw.fullscreen()
    # Center on screen and apply size
    screen_geometry = QApplication.desktop().screenGeometry()
    geometry = (
        args.geometry
        if args.geometry
        else parsertypes.Geometry(
            screen_geometry.width() / 2, screen_geometry.height() / 2
        )
    )
    x = (screen_geometry.width() - geometry.width) // 2
    y = (screen_geometry.height() - geometry.height) // 2
    mw.setGeometry(x, y, *geometry)


def update_settings(args):
    """Update default settings with command line arguments and configfiles.

    Args:
        args: Arguments returned from parser.parse_args().
    """
    configfile.parse(args)
    keyfile.parse(args)
    styles.parse()
    for option, value in args.cmd_settings:
        try:
            api.settings.override(option, value)
        except KeyError:
            logging.error("Unknown setting %s", option)
        except strconvert.ConversionError as e:
            logging.error(str(e))


def main():
    """Run startup and the Qt main loop."""
    qapp = app.Application()
    startup(sys.argv[1:])
    signal.signal(signal.SIGINT, signal.SIG_DFL)  # ^C
    returncode = qapp.exec_()
    plugins.cleanup()
    return returncode
